Refactoring all colors into a global Theme class and added basic support for themes.

GustavBjordal opened the pull request:
When working on the "profile compilation" feature last year I noticed that all colors are essentially magic constants spread all over the code and are sometimes used inconsistently. Furthermore, switching colors when going into dark mode was very messy and sometimes even overlooked.

I refactored all colors into a `Theme` class and added a global `currentTheme` object, all in a `Themes` namespace.
The `Theme` class forces all colors to be paired with a dark mode version of the color. This means that, for example, the current text color can be retrieved using `Themes::currentTheme.textColor.get(darkMode)`. Having a global object for the current theme is not great in my opinion, but it was the easiest approach that I could think of. Maybe this should be changed in the future.

I did these changes quite a while back (but forgot to create a pull request), so I may have missed some color constants in case they were added since then.

By doing this refactoring, it now becomes trivial to implement more themes for the syntax highlighting. I have pre-defined two other themes that can be selected via `View>Themes>` to show how this can be done. Feel free to remove/change this in case you disapprove of colors in general. I did the themes mainly for fun, the important part of this pull request is the refactoring. But I would be happy if they made it into the next release of course :-)

When I started doing this, I did some reading on what the standard approach for theming in Qt is. As far as I can tell there is none. It seemed like it is becoming increasingly popular to do theming _of UI elements_ in Qt using css, but there was no good way of doing _syntax highlighting_ via css. Furthermore, theming via css seems to require a massive effort as all UI elements seem to require explicit redefinitions (and one may need a better icon set). Finally, the behaviour of themes via css seems a bit weird since the OS seems to be able to override it (like macOS in dark mode).

zayenz commented on MiniZincIDE/highlighter.h:
```diff
@@ -22,6 +22,123 @@ struct Bracket {
+            ThemeColor(_commentColor,_commentColor),
+            ThemeColor(_backgroundColor,_backgroundColor),
+            ThemeColor(_lineHighlightColor,_lineHighlightColor),
+            ThemeColor(Qt::red,Qt::red))
```
Any reason this is not specifyable via a QColor?

GustavBjordal replied:
A Color is just a single color. The point of `ThemeColor` is to create a pair of colors, one for light and one for dark mode. Maybe this could have been done by extending the QColor class, but that seems like unnecessary work :)

zayenz replied:
I meant that this constructor can be used to set the theme color using a single QColor per color class, except for error color which is hard-coded.

GustavBjordal replied:
Ah yes sorry. 

Yeah I guess this is a bit strange. One can specify the error color using the contractor below, I just found that for my cases the error color is always red, so I made this constructor where this was just hard-coded so that I did not have to specify it. If this is too weird, we can just remove this constructor and use the one below.

By the way, going over the code now I see that there is actually no constructor for specifying all colors.
The colors `bracketsMatchColor`, `bracketsNoMatchColor`, `textHighlightColor`, `foregroundActiveColor`, and `foregroundInactiveColor` are derived from the other colors but the formulas are actually not safe for all input (safe in the sense that the colors can end up having bad contrasts). But adding the full constructor is trivial if ever required.

zayenz commented on MiniZincIDE/highlighter.h:
```diff
@@ -22,6 +22,123 @@ struct Bracket {
+            ThemeColor(Qt::red,Qt::red))
+            {}
+
+        Theme(ThemeColor _textColor,
```
While the calls constructing a Theme look ok in an IDE such as CLion that can annotate each argument with the parameter name, it is less clear in less capable IDEs or editors.

Unfortunately, C++ doesn't have keyword arguments, and none of the alternatives are really that great. So unclear to me if there are any really better ways to handle this.

GustavBjordal replied:
I agree on both points, but I cannot see a cleaner way of doing this.

The good news is that themes are rarely defined in the code, so not knowing which argument in the constructor is which is a rarely an issue. Also, C++20 does support constructors with named parameters, so this can easily be fixed in the future if the source migrates to C++20.

zayenz replied:
Yes, that will be a nice addition to have. And the calls in the code are well-documented, so not a large problem at all!

cyderize commented:
Hi Gustav, I've been playing around with this and unfortunately on Windows it doesn't seem to play nicely with the dark mode switch. For me the code editor background doesn't switch back to the light colour when disabling dark mode. I've been unable to figure out why so far, but hopefully we'll be able to fix it and get this into a future release.

GustavBjordal commented:
> Hi Gustav, I've been playing around with this and unfortunately on Windows it doesn't seem to play nicely with the dark mode switch. For me the code editor background doesn't switch back to the light colour when disabling dark mode. I've been unable to figure out why so far, but hopefully we'll be able to fix it and get this into a future release.

The issue seems to be with this code in mainwindow.cpp:
```
    if (darkMode) {
#ifndef Q_OS_MAC
        QFile sheet(":/dark_mode.css");
        sheet.open(QFile::ReadOnly);
        qApp->setStyleSheet(sheet.readAll());
#endif
    } else {
#ifndef Q_OS_MAC
        qApp->setStyleSheet("");
#endif
    }
```
I'm on macOS so I did not catch this.

I'll see if I can figure out what going on...

GustavBjordal commented:
I figured out what was wrong and fixed it.

The issue is that when stylesheets are used on a widget, the QPalette of that widget is disconnected. 
I had changed the styling of the codeditor to be done entirely using QPalette, but this would then break on non-macOS machines where the dark mode is done via stylesheets in mainwindow.cpp

The workaround is the go back to using some style sheeting for the code editor and generating the stylesheet on the fly using the theme colors (although it should be the other way around, where all styling is via QPalette on all platforms and css is not used anywhere).

The fix should now be part of the pull request.

cyderize commented:
Thanks, that seems to have fixed it for the code editor. I'm guessing a similar thing will need to be done for the output window too, since a the same thing seems to happen it it.

GustavBjordal commented:
Ah yes, I had missed that as well. Fixed now!

Actually it is a bit strange, `ui->outputConsole->document()->setDefaultStyleSheet(...)` does not interfere with `QPalette`, but the `qApp->setStyleSheet(...)` does. Qt is very confusing at times...

Note that the highlight color of the text in the output console (done via `ui->outputConsole->document()->setDefaultStyleSheet(...)`) does not update immediately when you switch between dark mode or between themes. This was also the case before this pull request. But any new text that is printed uses the correct color.

cyderize commented:
Awesome, I've rebased and merged this into develop in b1b254bcb2801239eb0f9d7c4fef46d21172a8a1

Thanks!
